feat: DIA-2148: write exports using thread pool

pakelley opened the pull request:
-------

<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

Optimize export job by threading writes to storage

pakelley committed:
feat: write exports using thread pool

pakelley commented on label_studio/io_storages/base_models.py:
```diff
@@ -551,15 +554,16 @@ def save_all_annotations(self):
-            self.info_update_progress(last_sync_count=annotation_exported, total_annotations=total_annotations)
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            futures = []
+            for annotation in Annotation.objects.filter(project=self.project).iterator(chunk_size=settings.STORAGE_EXPORT_CHUNK_SIZE):
```
notably, we have to submit all annots here before we update any progress, so I'm going to try either updating the counts in the thread or batching the updates (just using the same batch size that we use to fetch from the db, defaults to 100)

matt-bernstein commented on label_studio/io_storages/base_models.py:
```diff
@@ -551,15 +554,16 @@ def save_all_annotations(self):
-        ):
-            annotation.cached_user = self.cached_user
-            self.save_annotation(annotation)
+        max_workers = min(32, (os.cpu_count() or 4) * 4)
```
Do you see a need to let storage instances override this? (for different cloud providers, or in the future, provider-specific handling of max concurrent writes based on storage details) If so, it can be `self.max_workers`

pakelley committed:
batch writes

pakelley committed:
max 8 threads

pakelley committed:
make max_workers a class attribute

pakelley committed:
fix: don't accumulate anns across batches

pakelley committed:
Merge branch 'develop' into fb-dia-2148

pakelley commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14504694094)

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14504694094
